### PR TITLE
Launch cluster from Den with streaming logs

### PR DESCRIPTION
**Pull request by mkandler**
Allows for launching clusters via Den with streaming logs, WIP

### CODE REVIEW

**Comment by mkandler** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/run-house/runhouse/1343?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#1343** <a href="https://app.graphite.dev/github/pr/run-house/runhouse/1343?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#1336** <a href="https://app.graphite.dev/github/pr/run-house/runhouse/1336?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @mkandler and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: runhouse/resources/hardware/launcher_utils.py [hunk 1]
```diff
@@ -1,16 +1,21 @@
...
 from runhouse.constants import DOCKER_LOGIN_ENV_VARS
 from runhouse.globals import configs, rns_client
 from runhouse.logger import get_logger
+from runhouse.resources.hardware.utils import SSEClient
 from runhouse.rns.utils.api import load_resp_content
 
 logger = get_logger(__name__)
 
 
+class SSEHandler:
```
Comment (jlewitt1) [inline runhouse/resources/hardware/launcher_utils.py:14]: is this being used anywhere?

File: runhouse/resources/hardware/on_demand_cluster.py [hunk 2]
```diff
@@ -454,6 +454,19 @@ def _update_from_sky_status(self, dryrun: bool = False):
         cluster_dict = self._sky_status(refresh=not dryrun)
         self._populate_connection_from_status_dict(cluster_dict)
 
+    def _update_from_den_response(self, config: dict):
```
Comment (jlewitt1) [inline runhouse/resources/hardware/on_demand_cluster.py:457]: i think this should live in the DenLauncher class and be called at the end of the `up` method

File: runhouse/resources/hardware/launcher_utils.py [hunk 3]
```diff
@@ -155,8 +131,10 @@ def up(self, verbose: bool = True):
                     f"Received [{resp.status_code}] from Den POST '{self.LAUNCH_URL}': Failed to "
                     f"launch cluster: {load_resp_content(resp)}"
                 )
+            data = resp.json
 
         logger.info("Successfully launched cluster.")
+        return data
```
Comment (jlewitt1) [inline runhouse/resources/hardware/launcher_utils.py:137]: whats the data that we're returning here? `cluster.up` should return a cluster object, so in the case of the den launcher should return `self.cluster`

File: runhouse/resources/hardware/on_demand_cluster.py [hunk 4]
```diff
@@ -528,16 +541,27 @@ def up(self, verbose: bool = True, force: bool = False):
...
             logger.info("Launching cluster with Den.")
             launcher = DenLauncher(cluster=self, force=force)
+            data = launcher.up(verbose=verbose)
+            self._update_from_den_response(data)
+
         else:
             logger.info("Launching cluster locally.")
             launcher = LocalLauncher(cluster=self)
+            launcher.up(verbose=verbose)
+            self._update_from_sky_status()
```
Comment (jlewitt1) [inline runhouse/resources/hardware/on_demand_cluster.py:551]: we should probably call this inside the launcher too after the cluster has been upped and before returning to den

**Pull request merged**